packages/vm/src/testing/instance.rs

```diff
@@ -3,48 +3,107 @@
 
-use super::mock::{mock_dependencies, mock_dependencies_with_balances, MockApi, MockQuerier};
+use super::mock::{MockApi, MockQuerier, MOCK_CONTRACT_ADDR};
 use super::storage::MockStorage;
 
-/// Gas limit for testing
-const DEFAULT_GAS_LIMIT: u64 = 500_000;
-
-fn default_features() -> HashSet<String> {
-    HashSet::from_iter(["staking".to_string()].iter().cloned())
-}
-
 pub fn mock_instance(
     wasm: &[u8],
     contract_balance: &[Coin],
 ) -> Instance<MockStorage, MockApi, MockQuerier> {
-    check_wasm(wasm, &default_features()).unwrap();
-    let deps = mock_dependencies(20, contract_balance);
-    Instance::from_code(wasm, deps, DEFAULT_GAS_LIMIT).unwrap()
+    mock_instance_with_options(
+        wasm,
+        MockInstanceOptions {
+            contract_balance: Some(contract_balance),
+            ..Default::default()
+        },
+    )
 }
 
 pub fn mock_instance_with_balances(
     wasm: &[u8],
     balances: &[(&HumanAddr, &[Coin])],
 ) -> Instance<MockStorage, MockApi, MockQuerier> {
-    check_wasm(wasm, &default_features()).unwrap();
-    let deps = mock_dependencies_with_balances(20, balances);
-    Instance::from_code(wasm, deps, DEFAULT_GAS_LIMIT).unwrap()
+    mock_instance_with_options(
+        wasm,
+        MockInstanceOptions {
+            balances,
+            ..Default::default()
+        },
+    )
 }
 
 pub fn mock_instance_with_gas_limit(
     wasm: &[u8],
     contract_balance: &[Coin],
     gas_limit: u64,
 ) -> Instance<MockStorage, MockApi, MockQuerier> {
-    check_wasm(wasm, &default_features()).unwrap();
-    let deps = mock_dependencies(20, contract_balance);
-    Instance::from_code(wasm, deps, gas_limit).unwrap()
+    mock_instance_with_options(
+        wasm,
+        MockInstanceOptions {
+            contract_balance: Some(contract_balance),
+            gas_limit,
+            ..Default::default()
+        },
+    )
+}
+
+#[derive(Debug)]
+pub struct MockInstanceOptions<'a> {
+    // dependencies
+    pub canonical_address_length: usize,
+    pub balances: &'a [(&'a HumanAddr, &'a [Coin])],
+    /// This option is merged into balances and might override an existing value
+    pub contract_balance: Option<&'a [Coin]>,
+
+    // instance
+    pub supported_features: HashSet<String>,
+    pub gas_limit: u64,
+}
+
+impl Default for MockInstanceOptions<'_> {
+    fn default() -> Self {
+        Self {
+            // dependencies
+            canonical_address_length: 20,
+            balances: Default::default(),
+            contract_balance: Default::default(),
+
+            // instance
+            supported_features: features_from_csv("staking"),
+            gas_limit: 500_000,
+        }
+    }
+}
+
+pub fn mock_instance_with_options(
+    wasm: &[u8],
+    options: MockInstanceOptions,
+) -> Instance<MockStorage, MockApi, MockQuerier> {
+    check_wasm(wasm, &options.supported_features).unwrap();
+    let contract_address = HumanAddr::from(MOCK_CONTRACT_ADDR);
+
+    // merge balances
```

Comment:
Nice,
I wonder if we should store this in a HashMap internally?

Reply:
It becomes a map in the implementation of `MockQuerier::new`. I tried to use maps here as well but converting list -> map -> list with the correct lifetimes was too hard, so I have up.

However, we might decide to pass a map into `MockQuerier::new`.

Reply:
I don't have a strong opinion here. I'd merge as is to make it available for you now.

Reply:
Please merge as is.
It was just a passing thought, and didn't check that we did use a HashMap internally.